### ISSUE TITLE
how to Custom Barcode Arrangements?

### ISSUE THREAD

**Issue by SCQUchenyang** [nanoporetech/dorado#558]
Hi,
Can somebody help me please!
I used "PCB-109" kit, and followed by https://github.com/nanoporetech/dorado/blob/master/documentation/CustomBarcodes.md, I have set a barcode.toml to definte "LEADING_FLANK_1" and  " TRAILING_FLANK_1".The file is 
```
[arrangement]
name = "custom_barcode"
kit = "SQK-PCB109"
mask_1_front = "ATCGCCTACCGTGAC"
mask_1_rear = "ACTTGCCTGTCGCTCTATCTTCATTGATGGTGCCTACAG"
mask_2_front = ""
mask_2_rear = ""
# Barcode sequences
barcode1_pattern = "BC%02i"
barcode2_pattern = "BC%02i"
first_index = 1
last_index = 12
## Scoring options
[scoring]
min_soft_barcode_threshold = 0.2
min_hard_barcode_threshold = 0.2
min_soft_flank_threshold = 0.3
min_hard_barcode_threshold = 0.3
min_barcode_score_dist = 0.1
```
My command is ”dorado basecaller hac@v3.3output_pod5s/0 -x cuda:all --barcode-arrangement barcode.toml --emit-sam --estimate-poly-a >arrangement_barcode.sam “.

But Dorado failed to recognize the barcode for my reads. I want to know if there is an issue with my TOML file

**Comment by tijyojwad** [comment]
Hi @SCQUchenyang - this is an existing supported kit in dorado. you can use it with `--kit-name SQK-PCB109`. are you trying to update the thresholds with this custom config?

**Comment by SCQUchenyang** [comment]
> Hi @SCQUchenyang - this is an existing supported kit in dorado. you can use it with `--kit-name SQK-PCB109`. are you trying to update the thresholds with this custom config?
Hi, thanks for your suggestion, tijyojwad. Actually, there are flank squence (maybe named LEADING_FLANK_1 and TRAILING_FLANK_1 in https://github.com/nanoporetech/dorado/blob/master/documentation/CustomBarcodes.md) around the barcode, so I wrote them in the TOML file. But Dorado failed to recognize the barcode for my reads. 
And, I did not update the thresholds with this custom config.

**Comment by SCQUchenyang** [comment]
Hi, thanks for your suggestion, tijyojwad. Actually, there are flank squence (maybe named LEADING_FLANK_1 and TRAILING_FLANK_1 in https://github.com/nanoporetech/dorado/blob/master/documentation/CustomBarcodes.md) around the barcode, so I wrote them in the TOML file. But Dorado failed to recognize the barcode for my reads.
And, I did not update the thresholds with this custom config.

**Comment by tijyojwad** [comment]
Hi @SCQUchenyang - I think I'm missing something. Can you describe the experiment you're trying to run? i.e. what is the motivation for creating a custom config file for one of the pre-defined barcode kits from dorado?

**Comment by SCQUchenyang** [comment]
Hey, tijyojwad, thanks for your quick response. Let me rephrase this matter. Our laboratory has developed an experimental workflow (named FLEP-seq, https://doi.org/10.1038/s41596-021-00581-7) for detecting full-length mRNA. My reads should look like this example (in the picture), and we used barcode from SQK-PCB109. 
![image](https://github.com/nanoporetech/dorado/assets/37470873/8d2d8867-2f93-4421-8b6c-68ce86001838)

1.I want to detached polya length in my reads and I used with "--kit-name SQK-PCB109". The command like
```
dorado basecaller hac@v3.3 output_pod5s/0 -x cuda:all --kit-name SQK-PCB109 --emit-sam --estimate-poly-a >test.sam
```
Dorado successfully **identified the barcode in my reads, but did not detect the poly(A) length**. I have one million reads for test, but only 5000 of them were identified with a poly(A) tail.


2.And then, I realize that maybe “LEADING_FLANK_1” and “TRAILING_FLANK_1” have affected the results. As you can see, we have“LEADING_FLANK_1” and “TRAILING_FLANK_1” around the barcode, respectively. So, I have set a barcode.toml to definte "LEADING_FLANK_1" and " TRAILING_FLANK_1".The file is
```
[arrangement]
name = "custom_barcode"
kit = "SQK-PCB109"
mask_1_front = "ATCGCCTACCGTGAC"
mask_1_rear = "ACTTGCCTGTCGCTCTATCTTCATTGATGGTGCCTACAG"
mask_2_front = ""
mask_2_rear = ""
# Barcode sequences
barcode1_pattern = "BC%02i"
barcode2_pattern = "BC%02i"
first_index = 1
last_index = 12
## Scoring options
[scoring]
min_soft_barcode_threshold = 0.2
min_hard_barcode_threshold = 0.2
min_soft_flank_threshold = 0.3
min_hard_barcode_threshold = 0.3
min_barcode_score_dist = 0.1
```
The command is like
```
dorado basecaller hac@v3.3 output_pod5s/0 -x cuda:all --barcode-arrangement barcode.toml --emit-sam --estimate-poly-a >arrangement_barcode.sam
```
This time, Dorado successfully **detect the poly(A) length of reads** (300,000 out of 1,000,000 reads for test). But it doesn't identified the barcode in my reads. **No barcode information** was detected in the SAM file.
 

So, I want to know if there is an issue with my TOML file or my command?

**Comment by tijyojwad** [comment]
Hi @SCQUchenyang - I think I know what the problem might be. Potentially during barcode trimming we are getting rid of some of the primer that helps us locate the polyA. Can you try your run with `--no-trim` option? You don't need to provide the custom barcode arrangement file

**Comment by SCQUchenyang** [comment]
> --no-trim

Hi, tijyojwad.Thank you for your suggestion, I have already tried this. The command is like:
```
dorado basecaller hac@v3.3 output_pod5s/0 -x cuda:all --kit-name SQK-PCB109 --emit-sam --estimate-poly-a --no-trim >arrangement_barcode_0_notrim.sam
```
The results were not satisfactory. Only a very small percentage (less than 0.5%) of reads were identified with poly(A) tail. In fact, my test data should have  ~20% of poly(A) reads.
I feel that **a custom barcode arrangement file** might help me identify the poly(A) tail, but it doesn't identified the barcode. It seems that there might be an error in my TOML file.

**Comment by tijyojwad** [comment]
Hmm that's a bit surprising. Could you share a subset of the reads as a pod5 file? You can generate a subset using the pod5 tool - https://pod5-file-format.readthedocs.io/en/latest/docs/tools.html#pod5-filter

Just a few reads which you think have the barcode and polyA would be great.

There's another alternative I can suggest for now - you can split your run into 2 steps:

1. Poly estimation without barcode classification
```
dorado basecaller hac@v3.3 output_pod5s/0 -x cuda:all --emit-sam --no-trim --estimate-poly-a >test.sam
```

2. Then run barcode classification on that output. All tags should be retained
```
dorado demux --output-dir <output_dir> --no-trim --kit-name --kit-name SQK-PCB109 test.sam
```
This will output all barcode classified reads into the `<output_dir>` folder. You can also run without `--no-trim` if you don't mind the classified barcodes being trimmed.

**Comment by tijyojwad** [comment]
Hi @SCQUchenyang - I'm closing this thread due to inactivity. If you have more findings please re-open.
